### ISSUE TITLE
AllenNLP crf-tagger with BERT

### ISSUE THREAD

**Issue by Sneriko** [allenai/allennlp#4287]
Hi!

I tried to add swedish BERT embeddings to the AllenNLP crf-tagger, with the config file below, but got really bad results, even though the same BERT-model got really high F1-score when evaluating with the transformer standard model for sequence tagging. I was just wondering if anyone had any idea of why the results are so low. With BERT I got an F1-score of about 0.54, but with ELMO and FastText the score was about 0.90 (on the same dataset)

```
{
    "dataset_reader": {
        "type": "conll2003",
        "tag_label": "ner",
        "token_indexers": {
            "bert": {
                "type": "pretrained_transformer_mismatched",
                "model_name": "KB/bert-base-swedish-cased-ner"
            },
            "token_characters": {
                "type": "characters",
                "min_padding_length": 3
            }
        }
    },
    "model": {
        "type": "crf_tagger",
        "calculate_span_f1": true,
        "constrain_crf_decoding": true,
        "dropout": 0.5,
        "encoder": {
            "type": "lstm",
            "bidirectional": true,
            "dropout": 0.5,
            "hidden_size": 200,
            "input_size": 896,
            "num_layers": 2
        },
        "include_start_end_transitions": false,
        "label_encoding": "IOB1",
        "text_field_embedder": {
            "token_embedders": {
                "bert": {
                    "type": "pretrained_transformer_mismatched",
                    "model_name": "KB/bert-base-swedish-cased-ner"
                },
                "token_characters": {
                    "type": "character_encoding",
                    "embedding": {
                        "embedding_dim": 16,
                        "vocab_namespace": "token_characters"
                    },
                    "encoder": {
                        "type": "cnn",
                        "conv_layer_activation": "relu",
                        "embedding_dim": 16,
                        "ngram_filter_sizes": [
                            3
                        ],
                        "num_filters": 128
                    }
                }
            }
        }
    },
    "train_data_path": "Data/Datasets/ner/SUC3trainCONLL2003",
    "validation_data_path": "Data/Datasets/ner/SUC3testCONLL2003",
    "test_data_path": "Data/Datasets/ner/SUC3testCONLL2003",
    "trainer": {
        "cuda_device": 0,
        "grad_norm": 5,
        "num_epochs": 20,
        "optimizer": {
            "type": "adam",
            "lr": 0.001
        },
        "patience": 5,
        "validation_metric": "+f1-measure-overall"
    },
    "data_loader": "allennlp.data.dataloader.DataLoader"
}
```

**Comment by epwalsh** [comment]
@Sneriko that's interesting. Could you also share your config that you used with ELMO so that we have something to compare this with?

**Comment by Sneriko** [comment]
yes, absolutely.

{
    "dataset_reader": {
        "type": "conll2003",
        "coding_scheme": "BIOUL",
        "tag_label": "ner",
        "token_indexers": {
            "elmo": {
                "type": "elmo_characters"
            },
            "token_characters": {
                "type": "characters",
                "min_padding_length": 3
            },
            "tokens": {
                "type": "single_id",
                "lowercase_tokens": true
            }
        }
    },
    "model": {
        "type": "crf_tagger",
        "dropout": 0.5,
        "encoder": {
            "type": "lstm",
            "bidirectional": true,
            "dropout": 0.5,
            "hidden_size": 200,
            "input_size": 1452,
            "num_layers": 2
        },
        "include_start_end_transitions": false,
        "label_encoding": "BIOUL",
        "regularizer": {
            "regexes": [
                [
                    "scalar_parameters",
                    {
                        "alpha": 0.1,
                        "type": "l2"
                    }
                ]
            ]
        },
        "text_field_embedder": {
            "token_embedders": {
                "elmo": {
                    "type": "elmo_token_embedder",
                    "do_layer_norm": false,
                    "dropout": 0,
                    "options_file": "Data/elmo/swedish/options.json",
                    "weight_file": "Data/elmo/swedish/swedish-elmo-weights.hdf5"
                },
                "token_characters": {
                    "type": "character_encoding",
                    "embedding": {
                        "embedding_dim": 16,
                        "vocab_namespace": "token_characters"
                    },
                    "encoder": {
                        "type": "cnn",
                        "conv_layer_activation": "relu",
                        "embedding_dim": 16,
                        "ngram_filter_sizes": [
                            3
                        ],
                        "num_filters": 128
                    }
                },
                "tokens": {
                    "type": "embedding",
                    "embedding_dim": 300,
                    "pretrained_file": "Data/FastTextVectors/cc.sv.300.vec",
                    "trainable": true
                }
            }
        }
    },
    "train_data_path": "Data/Datasets/ner/SUC3trainCONLL2003",
    "validation_data_path": "Data/Datasets/ner/SUC3testCONLL2003",
    "test_data_path": "Data/Datasets/ner/SUC3testCONLL2003",
    "trainer": {
        "cuda_device": 0,
        "grad_norm": 5,
        "num_epochs": 35,
        "optimizer": {
            "type": "adam",
            "lr": 0.001
        },
        "patience": 8,
        "validation_metric": "+f1-measure-overall"
    },
    "vocabulary": {
        "min_pretrained_embeddings": {
            "tokens": 200000
        },
        "pretrained_files": {
            "tokens": "Data/FastTextVectors/cc.sv.300.vec"
        }
    },
    "data_loader": "allennlp.data.dataloader.DataLoader",
    "evaluate_on_test": true
}

**Comment by epwalsh** [comment]
Thanks @Sneriko, and could also tell me which version of AllenNLP you used? Or better yet, just paste the output of `pip freeze`.

**Comment by epwalsh** [comment]
I see a couple of differences between your two configs that stand out to me:
- In your BERT config, you are using `IOB1` coding scheme (the default), whereas in your ELMO config you're suing `BIOUL`.
- You're using a regularizer in the ELMO config but not the BERT config.

I don't think any of these differences on their own are enough to explain the performance gap, but it may be worth rectifying these differences first to see if you still get the same results.

Also, when you say

> With BERT I got an F1-score of about 0.54, but with ELMO and FastText the score was about 0.90 (on the same dataset)

Are you talking about the test set or the validation set? And is this a public dataset?

**Comment by epwalsh** [comment]
You might want to have a look at this config: https://github.com/allenai/allennlp-models/blob/master/training_config/rc/transformer_qa.jsonnet

This could be a good starting point as far as learning rate and other hyper parameters for fine-tuning a BERT model.

**Comment by edchengg** [comment]
You might want to try downgrade allennlp to 0.9. The current 1.0 does not use transformers lib in a correct way. I encounter similar issue #4392

**Comment by matt-gardner** [comment]
If there is an issue with the BERT SRL model, it's isolated to the issue in #4392. The issue here is very different. We have plenty of models that train correctly with the transformers library.
